### PR TITLE
docs(config): add godoc for the config package

### PR DESCRIPTION
**Pull request by Joshua-Anderson**

### CODE REVIEW

File: api/config.go [hunk 1]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
 // Config is the structure of an app's config.
 type Config struct {
-	Owner       string                  `json:"owner,omitempty"`
-	App         string                  `json:"app,omitempty"`
-	Values      map[string]interface{}  `json:"values,omitempty"`
-	Memory      map[string]interface{}  `json:"memory,omitempty"`
-	CPU         map[string]interface{}  `json:"cpu,omitempty"`
+	// Owner is the app owner. It cannot be updated with config.Set(). See app.Transfer().
+	Owner string `json:"owner,omitempty"`
+	// App is the app name. It cannot be update at all at this moment.
```
Comment (mboersma) [inline api/config.go]: Typo, should be "update**d**".

File: api/config.go [hunk 2]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
+	App string `json:"app,omitempty"`
+	// Values are exposed as environmantal variables to the app.
```
Comment (mboersma) [inline api/config.go]: Typo: "environmental"

File: api/config.go [hunk 3]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
+	// Healthchecks are the healthchecks that the application uses.
 	Healthcheck map[string]*Healthcheck `json:"healthcheck,omitempty"`
-	Tags        map[string]interface{}  `json:"tags,omitempty"`
-	Registry    map[string]interface{}  `json:"registry,omitempty"`
-	Created     string                  `json:"created,omitempty"`
-	Updated     string                  `json:"updated,omitempty"`
-	UUID        string                  `json:"uuid,omitempty"`
+	// Tags restrict applications to run on k8s nodes with that label.
+	Tags map[string]interface{} `json:"tags,omitempty"`
+	// Registry is a key value pair to provide authentication for docker registries.
```
Comment (mboersma) [inline api/config.go]: I think "key**-**value pair" with a hyphen is more common.

File: api/config.go [hunk 4]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
+	// The key is the username and the value is the password.
+	Registry map[string]interface{} `json:"registry,omitempty"`
+	// Created is the time that the applications was created and cannot be updated.
```
Comment (mboersma) [inline api/config.go]: Typo: "application"

File: api/config.go [hunk 5]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
-	CPU         map[string]interface{}  `json:"cpu,omitempty"`
+	// Owner is the app owner. It cannot be updated with config.Set(). See app.Transfer().
+	Owner string `json:"owner,omitempty"`
+	// App is the app name. It cannot be update at all at this moment.
+	App string `json:"app,omitempty"`
+	// Values are exposed as environmantal variables to the app.
+	Values map[string]interface{} `json:"values,omitempty"`
+	// Memory is used to set process memory limits. The key is the process name
+	// and the value is a number followed by a memory unit (G, M, K, or B). Ex: 200G
+	Memory map[string]interface{} `json:"memory,omitempty"`
```
Comment (arschles) [inline api/config.go]: is it possible to make the `interface{}`s into more concrete types?

File: api/config.go [hunk 6]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
+	// Tags restrict applications to run on k8s nodes with that label.
+	Tags map[string]interface{} `json:"tags,omitempty"`
+	// Registry is a key value pair to provide authentication for docker registries.
+	// The key is the username and the value is the password.
+	Registry map[string]interface{} `json:"registry,omitempty"`
+	// Created is the time that the applications was created and cannot be updated.
+	Created string `json:"created,omitempty"`
+	// Updated is the last time the configuration was changed and cannot be updated.
+	Updated string `json:"updated,omitempty"`
+	// UUID is a unique string reflecting the configuration in it's current state.
```
Comment (mboersma) [inline api/config.go]: Typo: "its".

File: api/config.go [hunk 7]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
+	// It changes every time the configuration is changed and cannot be updated
```
Comment (mboersma) [inline api/config.go]: Missing a final period.

Nitpick level up! 💥

File: api/config.go [hunk 8]
```diff
@@ -56,9 +71,13 @@ Failure Threshold: {{.FailureThreshold}}
 Exec Probe: {{or .Exec "N/A"}}
 HTTP GET Probe: {{or .HTTPGet "N/A"}}
 TCP Socket Probe: {{or .TCPSocket "N/A"}}`)
-	if err != nil { panic(err) }
+	if err != nil {
+		panic(err)
```
Comment (arschles) [inline api/config.go]: is it possible to not panic in the SDK, and instead return errors?

Comment (arschles) [inline api/config.go]: not blocking for this PR

File: api/config.go [hunk 9]
```diff
@@ -56,9 +71,13 @@ Failure Threshold: {{.FailureThreshold}}
...
+	}
 	err = tmpl.Execute(&doc, h)
```
Comment (mboersma) [inline api/config.go]: Super nit-pick, but we seem to prefer the one-liner version of this Go construct when possible:
```go
if err = tmpl.Execute(&doc, h); err != nil {
        panic(err)
}
```

File: config/config.go [hunk 10]
```diff
@@ -32,7 +33,17 @@ func List(c *deis.Client, app string) (api.Config, error) {
...
+// This is a patching operation, which means when you call Set() with an api.Config:
+//
+//    - If the variable does not exist, it will be set.
+//    - If the variable exists, it will be overwriten.
+//    - If the variable is set to nil, it will be unset.
+//    - If the variable was ignored in the api.Config, it will remain unchanged.
+//
+// Calling Set() with an empty api.Config will return a deis.ErrConflict.
+// Trying to unset a key that does not exist returns a deis.ErrUnprocessable.
+// Trying to set a tag that is not a label in the kubernetes cluster will return a deis.ErrTagNotFound
```
Comment (mboersma) [inline config/config.go]: Missing a final period.

File: api/config.go (hunk 8 above)
Comment (Joshua-Anderson) [inline api/config.go]: IDK @bacongobbler wrote it pretty recently. My editor just runs gofmt whenever I save and I guess this got caught :)

File: api/config.go (hunk 9 above)
Comment (arschles) [inline api/config.go]: I'd amend that to define the error only inside the block (notice the `err :=`)

```go
if err := tmpl.Execute(&doc, h); err != nil {
        panic(err)
}
```

File: api/config.go (hunk 5 above)
Comment (Joshua-Anderson) [inline api/config.go]: I'm interested in doing so, but it's out of the score of this PR. We should open an issue on it.

File: api/config.go (hunk 9 above)
Comment (Joshua-Anderson) [inline api/config.go]: I think this is out of scope for doc PR. I'll fix it up in a subsequent PR. I guess my auto-formatter just caught it by default.

**Comment by Joshua-Anderson** [comment]
Thanks for the feedback, I fixed up all the typos.

File: api/config.go [hunk 11]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
-	Owner       string                  `json:"owner,omitempty"`
-	App         string                  `json:"app,omitempty"`
-	Values      map[string]interface{}  `json:"values,omitempty"`
-	Memory      map[string]interface{}  `json:"memory,omitempty"`
-	CPU         map[string]interface{}  `json:"cpu,omitempty"`
+	// Owner is the app owner. It cannot be updated with config.Set(). See app.Transfer().
+	Owner string `json:"owner,omitempty"`
+	// App is the app name. It cannot be updated at all right now.
+	App string `json:"app,omitempty"`
+	// Values are exposed as environmental variables to the app.
```
Comment (bacongobbler) [inline api/config.go]: environment*

File: api/config.go (hunk 8 above)
Comment (bacongobbler) [inline api/config.go]: I think it's fine to panic if the SDK cannot compile the template, rather I think this is a valid situation where a panic is warranted. There is no way someone could recover from this, and a `String()` method returns no errors, anyways.

File: api/config.go (hunk 9 above)
Comment (bacongobbler) [inline api/config.go]: my bad, thanks for catching this!

File: api/config.go [hunk 12]
```diff
@@ -18,17 +18,32 @@ type ConfigUnset struct {
...
+	Values map[string]interface{} `json:"values,omitempty"`
+	// Memory is used to set process memory limits. The key is the process name
+	// and the value is a number followed by a memory unit (G, M, K, or B). Ex: 200G
+	Memory map[string]interface{} `json:"memory,omitempty"`
+	// CPU is used to set process CPU limits. The can be set in terms of whole CPUs
```
Comment (bacongobbler) [inline api/config.go]: "The can"? :)

**Comment by Joshua-Anderson** [comment]
@bacongobbler I fixed the issues you found!

**Pull request merged**